File: src/ops.js
// Op capture: turning a committed document mutation into the record-level
// put/delete list that the journal persists.
//
// An op names a map and a record id and either assigns the full record value
// or deletes it. Records are the unit — not fields — so a put always carries
// the whole record. That is deliberate: the server can apply and compact ops
// with no idea what an event or a frame is.
//
// Two capture strategies feed the same op shape:
//
//   * Bundle diff, for the transaction helpers. Those already collect the
//     before/after records surrounding an edit, and that bundle IS the set of
//     records the edit may touch, so every record in the after-bundle becomes
//     a put and anything that vanished becomes a delete.
//
//   * Reference diff, for wholesale rewrites like an ICS import or calendar
//     pull, where deep-copying the document to compare it would cost more than
//     the edit itself. A shallow snapshot of each map is taken before the
//     mutation and compared by object identity afterwards: a record that still
//     points at the same object was not rewritten. This is only sound because
//     those paths replace whole records instead of mutating them in place —
//     see the invariant noted in `src/calendar-sync.js`.
//
// Op values hold live references into the document rather than copies. The
// store serializes them when it posts, so a record edited again inside the
// same debounce window is persisted once, in its final state.

export const RECORD_MAPS = ["events", "frames", "patterns", "relations", "overrides", "meta", "foreign"];

const RECORD_MAP_SET = new Set(RECORD_MAPS);

export function putOp(map, id, value) {
  return { op: "put", map, id, value };
}

export function delOp(map, id) {
  return { op: "del", map, id };
}

// Applies record-level ops to a document in place. This is the one
// implementation of what an op MEANS: the client replays a conflicting
// window's ops with it, and the server replays the journal and compacts with
// it. Two copies of this that drifted apart would corrupt documents quietly,
// so `tools/journal.js` imports this rather than keeping its own.
//
// Ops are idempotent by construction — applying the same put or delete twice
// lands on the same state — which is what makes a crash mid-compaction
// survivable and makes conflict rebasing safe to retry.
export function applyOps(document, ops = []) {
  let applied = 0;
  for (const op of ops) {
    if (!op || typeof op !== "object") throw new TypeError("journal op must be an object");
    if (!RECORD_MAP_SET.has(op.map)) throw new RangeError(`unknown journal map "${op.map}"`);
    if (typeof op.id !== "string" || !op.id) throw new TypeError("journal op needs a record id");
    const map = (document[op.map] ||= {});
    if (op.op === "put") {
      if (op.value === undefined) throw new TypeError(`put op for ${op.map}/${op.id} carries no value`);
      map[op.id] = op.value;
    } else if (op.op === "del") {
      delete map[op.id];
    } else {
      throw new RangeError(`unknown journal op "${op.op}"`);
    }
    applied += 1;
  }
  return applied;
}

// A shallow, per-map snapshot of record identities. Cheap by construction: one
// pointer per record, no traversal into the records themselves.
export function mapSnapshot(document, maps = RECORD_MAPS) {
  const snapshot = {};
  for (const map of maps) snapshot[map] = { ...(document[map] || {}) };
  return snapshot;
}

// Diffs two per-map record collections into ops. A record present in `after`
// under the same identity as in `before` is treated as untouched; anything
// else is a put, and anything that disappeared is a delete.
export function opsFromMaps(before, after) {
  const ops = [];
  for (const map of RECORD_MAPS) {
    const previous = before[map] || {};
    const next = after[map] || {};
    for (const [id, value] of Object.entries(next)) {
      if (id in previous && previous[id] === value) continue;
      ops.push(putOp(map, id, value));
    }
    for (const id of Object.keys(previous)) {
      if (!(id in next)) ops.push(delOp(map, id));
    }
  }
  return ops;
}

// Normalizes the several bundle shapes the transaction helpers use into the
// per-map form `opsFromMaps` expects. `captureEventBundle` carries one `event`
// plus related maps; `captureFrameBundle` carries one `frame`; the set variants
// already carry maps.
export function bundleMaps(bundle, { eventId = null, frameId = null } = {}) {
  const maps = { events: {}, frames: {}, patterns: {}, relations: {}, overrides: {} };
  if (!bundle) return maps;
  if (bundle.events) Object.assign(maps.events, bundle.events);
  if (bundle.frames) Object.assign(maps.frames, bundle.frames);
  if (bundle.relations) Object.assign(maps.relations, bundle.relations);
  if (bundle.patterns) Object.assign(maps.patterns, bundle.patterns);
  if (bundle.overrides) Object.assign(maps.overrides, bundle.overrides);
  if (eventId && bundle.event) maps.events[eventId] = bundle.event;
  if (frameId && bundle.frame) maps.frames[frameId] = bundle.frame;
  return maps;
}

// Forward and inverse ops for one edit, from its before/after bundles. The
// inverse is what undo commits: the bundles hold the before-values already, so
// no extra capture is needed.
export function bundleOps(before, after, scope = {}) {
  const beforeMaps = bundleMaps(before, scope);
  const afterMaps = bundleMaps(after, scope);
  return {
    ops: opsFromMaps(beforeMaps, afterMaps),
    inverseOps: opsFromMaps(afterMaps, beforeMaps)
  };
}

// Forward and inverse ops for a single record in one map.
export function recordOps(map, id, before, after) {
  return {
    ops: after === undefined ? [delOp(map, id)] : [putOp(map, id, after)],
    inverseOps: before === undefined ? [delOp(map, id)] : [putOp(map, id, before)]
  };
}

// Accumulates committed edits until the store posts them. One entry per
// commit, so the debounce batches several edits into a single request without
// flattening them into one indistinguishable blob — the journal keeps each
// edit's own label and op list. Order is preserved, because a put followed by
// a delete of the same record means something different than the reverse.
export class OpLog {
  constructor() {
    this.entries = [];
  }

  get length() {
    return this.entries.length;
  }

  get opCount() {
    return this.entries.reduce((total, entry) => total + entry.ops.length, 0);
  }

  collect(label, ops = []) {
    if (!ops?.length) return this;
    this.entries.push({ label, ops });
    return this;
  }

  // Returns the pending entries and empties the log. A failed post hands them
  // back with `restore` so nothing is lost.
  drain() {
    const entries = this.entries;
    this.entries = [];
    return entries;
  }

  restore(entries = []) {
    if (entries.length) this.entries = [...entries, ...this.entries];
    return this;
  }

  clear() {
    this.entries = [];
  }
}

File: app/lib/core/ops.dart
// The op vocabulary: what a committed change to the document IS, as the record
// list a journal can persist and replay.
//
// An op names a map and a record id and either assigns the WHOLE record or
// deletes it. Records are the unit, never fields. That is deliberate: a journal
// can apply and compact ops with no idea what an event or a frame is, and
// applying the same put or delete twice lands on the same state -- which is what
// makes a crash mid-write survivable and a retry safe.
//
// THERE IS EXACTLY ONE DEFINITION HERE. The JavaScript kept a second copy of
// `applyOps` in its journal tool; two copies that drifted apart would corrupt
// documents quietly, so [applyOps] below is the only one that exists.

import 'package:freezed_annotation/freezed_annotation.dart';

import 'records.dart';

part 'ops.freezed.dart';

/// One record-level op. `value` holds either a live record or the parsed JSON a
/// journal replay hands back, so a put carries the whole record either way.
/// `op` is `put` or `del`, read by [applyOps] and by nothing else.
@Freezed(fromJson: false, toJson: false)
abstract class Op with _$Op {
  const Op._();
  const factory Op({required String op, required String map, required String id, Object? value}) =
      _Op;

  factory Op.fromJson(Json json) => Op(
    op: '${json['op'] ?? ''}',
    map: '${json['map'] ?? ''}',
    id: '${json['id'] ?? ''}',
    value: json['value'],
  );

  Json toJson() => {'op': op, 'map': map, 'id': id, if (value != null) 'value': jsonValue(value)};
}

Op putOp(String map, String id, Object? value) => Op(op: 'put', map: map, id: id, value: value);

Op delOp(String map, String id) => Op(op: 'del', map: map, id: id);

/// The one definition of what an op means. Idempotent by construction, and
/// total: an unknown map or verb is refused by name rather than skipped, because
/// a silently dropped op is a document that quietly disagrees with its journal.
Document applyOps(Document document, Iterable<Op> ops) {
  var next = document;
  for (final op in ops) {
    if (op.id.isEmpty) throw ArgumentError('journal op needs a record id');
    next = switch (op.op) {
      'put' => next.put(op.map, op.id, op.value),
      'del' => next.remove(op.map, op.id),
      _ => throw ArgumentError('unknown journal op "${op.op}"'),
    };
  }
  return next;
}

/// A per-map snapshot of record identities, taken before an edit. Cheap by
/// construction: one reference per record, no traversal into the records.
Map<String, Map<String, Object?>> mapSnapshot(
  Document document, [
  List<String> maps = recordMaps,
]) => {for (final map in maps) map: Map.of(document.records(map))};

/// The identity diff. A record that is still the same value was not rewritten,
/// so it emits nothing; anything else is a put, and anything that vanished is a
/// delete. Immutable records make the precondition structural -- no in-place
/// mutation exists that could leave a changed record looking untouched -- which
/// is why this is the whole capture strategy rather than a fallback for one.
List<Op> opsFromMaps(
  Map<String, Map<String, Object?>> before,
  Map<String, Map<String, Object?>> after,
) {
  final ops = <Op>[];
  for (final map in recordMaps) {
    final previous = before[map] ?? const {};
    final next = after[map] ?? const {};
    for (final entry in next.entries) {
      final was = previous[entry.key];
      final untouched = identical(was, entry.value) || was == entry.value;
      if (previous.containsKey(entry.key) && untouched) continue;
      ops.add(putOp(map, entry.key, entry.value));
    }
    for (final id in previous.keys) {
      if (!next.containsKey(id)) ops.add(delOp(map, id));
    }
  }
  return ops;
}

/// Forward and inverse ops for one edit. ONE bundle shape: the three the
/// JavaScript carried (an event plus its related maps, a frame plus its related
/// maps, and the already-per-map set variants) were the same per-map collection
/// wearing three constructors, and the inverse needs no extra capture because
/// the before-collection already holds it.
({List<Op> ops, List<Op> inverseOps}) bundleOps(
  Map<String, Map<String, Object?>> before,
  Map<String, Map<String, Object?>> after,
) => (ops: opsFromMaps(before, after), inverseOps: opsFromMaps(after, before));

/// One committed edit's label and ops.
typedef OpEntry = ({String label, List<Op> ops});

/// Accumulates committed edits until the store writes them. One entry per
/// commit, so a debounce batches several edits into one write without flattening
/// them into an indistinguishable blob. Order is preserved: a put followed by a
/// delete of the same record means something different than the reverse.
class OpLog {
  final List<OpEntry> _entries = [];

  List<OpEntry> get entries => List.unmodifiable(_entries);

  int get length => _entries.length;

  int get opCount => _entries.fold(0, (total, entry) => total + entry.ops.length);

  void collect(String label, List<Op> ops) {
    if (ops.isNotEmpty) _entries.add((label: label, ops: ops));
  }

  /// The pending entries, emptying the log. A failed write hands them back with
  /// [restore] so nothing is lost.
  List<OpEntry> drain() {
    final drained = List.of(_entries);
    _entries.clear();
    return drained;
  }

  void restore(List<OpEntry> entries) => _entries.insertAll(0, entries);

  void clear() => _entries.clear();
}
